File: prodplus-api/src/main/java/br/com/prodplus/models/Custo.java
package br.com.prodplus.models;

import java.io.Serializable;
import java.math.BigDecimal;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.PositiveOrZero;

import br.com.prodplus.models.enums.Periodo;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

/**
 * 
 * @author Marlon Fernando Garcia
 *
 */
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Entity
public class Custo implements Serializable {

	private static final long serialVersionUID = -7203875387573979548L;
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Integer id;
	@Column(nullable = false, unique = true)
	@NotBlank(message = "a descrição é obrigatória!")
	private String descricao;
	@Enumerated(EnumType.STRING)
	@Column(nullable = false, length = 20)
	@NotNull(message = "o período é obrigatório!")
	private Periodo periodo;
	@Column(nullable = false)
	@NotNull(message = "o valor é obrigatório!")
	@PositiveOrZero(message = "valor inválido!")
	private BigDecimal valor;
	@Column(nullable = false)
	@NotNull(message = "campo obrigatório")
	private boolean ativo = true;

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((descricao == null) ? 0 : descricao.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Custo other = (Custo) obj;
		if (descricao == null) {
			if (other.descricao != null)
				return false;
		} else if (!descricao.equals(other.descricao))
			return false;
		return true;
	}

}

File: prodplus/src/app/models/custo.ts
export class Custo {
  id: number | null;
  descricao: string;
  periodo: string;
  valor: number;
  ativo: boolean;

  constructor();
  constructor(
    descricao: string,
    periodo: string,
    valor: number,
    ativo: boolean,
    id?: number
  );
  constructor(...args: any[]) {
    if (args.length == 0) {
      this.id = null;
      this.descricao = '';
      this.periodo = '';
      this.valor = 0;
      this.ativo = true;
    } else {
      this.descricao = args[0];
      this.periodo = args[1];
      this.valor = args[2];
      this.ativo = args[3];
      if (typeof args[4] === 'undefined') this.id = null;
      else this.id = args[4];
    }
  }
}
